src/CameraSensor.cc

```diff
@@ -254,6 +254,20 @@ bool CameraSensor::CreateCamera()
     cameraSdf->SetLensIntrinsicsCy(intrinsicMatrix(1, 2));
   }
 
+  if(!cameraSdf->HasLensProjection())
+  {
+    auto intrinsicMatrix =
+      gz::rendering::projectionToCameraIntrinsic(
+        this->dataPtr->camera->ProjectionMatrix(),
+        this->dataPtr->camera->ImageWidth(),
+        this->dataPtr->camera->ImageHeight()
+      );
+    cameraSdf->SetLensProjectionFx(intrinsicMatrix(0, 0));
+    cameraSdf->SetLensProjectionFy(intrinsicMatrix(1, 1));
+    cameraSdf->SetLensProjectionCx(intrinsicMatrix(0, 2));
+    cameraSdf->SetLensProjectionCy(intrinsicMatrix(1, 2));
+  }
```

Comment:
remove this duplicate block? (same code as lines above)

Reply:
the code above is setting the intrinsic parameters this one is setting the projection parameters

Reply:
ahh  I see now :see_no_evil:

Reply:
I'm also confused about this. Is the conversion from projection to intrinsic matrix necessary here?

Reply:
This method get these values https://github.com/gazebosim/gz-rendering/blob/c70882b96fc126e13fa82bc60ead17bf803bec08/include/gz/rendering/base/BaseCamera.hh#L505-L522

Maybe there is another way. @iche033 ?

Reply:
@adityapande-1995 might also be interested.

Reply:
Yeah I agree, right now I think this section is writing the intrinsics matrix into to projection matrix section in SDF.

Maybe it should be : 
```
  if(!cameraSdf->HasLensProjection())
  {
    auto projectionMatrix = this->dataPtr->camera->ProjectionMatrix();
    
    cameraSdf->SetLensProjectionFx(projectionMatrix(0, 0));
    cameraSdf->SetLensProjectionFy(projectionMatrix(1, 1));
    cameraSdf->SetLensProjectionCx(projectionMatrix(0, 2));
    cameraSdf->SetLensProjectionCy(projectionMatrix(1, 2));
  } 
```

Reply:
I tried these changes and it's not working, I think we should stick with the current state of the PR

Reply:
I think that's because the `ProjectionMatrix()` returned here is for use in computer graphics / OpenGL applications and that's slightly different from what computer vision /OpenCV uses. 

Btw, just looking at the code above, it seems like camera projection matrix == intrinsic matrix in SDF? Since both elements in SDF are set to the same intrinsics matrix returned by gz-rendering.

Reply:
OpenGL projection matrix is:

![Selection_101](https://user-images.githubusercontent.com/1933907/203626265-e2f0a447-6979-4f96-a5ac-547a2094db87.png)

Reply:
@azeey  and @iche033  https://github.com/gazebosim/sdformat/pull/1088#discussion_r927942615

Reply:
accoring with `camera.sdf` , these values override the others

```xml
      <element name="p_fx" type="double" default="277" required="0">
        <description>X focal length for projection matrix(in pixels, overrides fx)</description>
      </element>
      <element name="p_fy" type="double" default="277" required="0">
        <description>Y focal length for projection matrix(in pixels, overrides fy)</description>
      </element>
      <element name="p_cx" type="double" default="160" required="0">
        <description>X principal point for projection matrix(in pixels, overrides cx)</description>
      </element>
      <element name="p_cy" type="double" default="120" required="0">
        <description>Y principal point for projection matrix(in pixels, overrides cy)</description>
      </element>
      <element name="tx" type="double" default="0.0" required="0">
        <description>X translation for projection matrix (in pixels)</description>
      </element>
```

Reply:
cool, I think as long as it follows the spec it should be fine. Can you add a brief comment here just to mention this? ie the projection matrix overrides intrinsic matrix according to sdf spec? thanks.

Reply:
From my understanding reading the description of the SDFormat `projection` element, it's only necessary to set the projection matrix to override what would normally be computed using the intrinsic matrix and other parameters of the camera. If we don't need to override it, there should be no need to set the projection matrix.
So the question is why are we overriding the intrinsic matrix here?

> Camera projection matrix P for overriding camera intrinsic matrix K values so that users can configure P independently of K. This is useful when working with stereo cameras where P may be different from K due to the transform between the two cameras.

@ahcorde In the PR description, you mentioned "This PR solves the camera projection matrix values not being reflected properly in /camera_info topic when the tag is not provided. " Can you elaborate this a little more? What's the value you expect in `/camera_info` and what are you getting instead? Is there an example or issue I can look at to find out more about the problem?

Reply:
I think I understand a little better now. But it seems to me that if the original SDF didn't have `<projection>` set, then the message should also not set the `Projection` field.
Would checking if `cameraSdf->HasLensProjection()` in `CameraSensor::PopulateInfo` and leaving it unpopulated in the `camera_info.proto` be a solution? https://github.com/gazebosim/gz-sensors/blob/c455a36f760559bdf945eee61b2fb446a97bba7b/src/CameraSensor.cc#L671-L673

Reply:
You can create the projection matrix based on the intrinsic matrix and camera pose, I would perfer not to leave this unpopulated. This might break users

Reply:
Okay, but note that there wouldn't be any way for users to know if the projection values were set by the user to override the values in `<intrinsic>` or if they were computed based on what's in `<intrinsic>`. If that's not important, then I'm good with this change. Since this is all very confusing, I would ask that you would add a detailed comment explaining why we're doing this and why the `projectionToCameraIntrinsic` is being used (i.e how Ogre2 projection matrix is not the same as the projection matrix from `<projection>`).

Reply:
@azeey  I added a comment here https://github.com/gazebosim/gz-sensors/pull/289/commits/f4b77537a2ad16ba8f83cf971aa416aaab6fd6d1

Reply:
Thanks! I would suggest adding "Note that the matrix from Ogre via `camera->ProjectionMatrix()` has a different format than the projection matrix used in SDFormat. This is why they are converted using `projectionToCameraIntrinsic`. The resulting matrix is the intrinsic matrix, but since the user has not overridden the values, this is also equal to the projection matrix"

That's my understanding at least. If that's not correct, feel free to modify it.

Reply:
@azeey Added here https://github.com/gazebosim/gz-sensors/pull/289/commits/32a6241cf63816929fec1b99f11e78e277341f14

Reply:
LGTM! Thanks